sympy/core/function.py

```diff
@@ -1213,6 +1213,10 @@ def __new__(cls, expr, *variables, **kwargs):
         # derivative class by calling Expr.__new__.
         if (not (hasattr(expr, '_eval_derivative') and evaluate) and
```

Comment:
I guess this logic was already here, but this seems like a bit of a code smell to me. I think this logic would be better implemented as a default method on the base class, and a custom method on `Derivative`. Then Python's inheritance would take care of which logic gets used automatically. It also makes it easy to extend the logic because a subclass could implement `_eval_derivative` and still call the superclass's version. 

This also clarifies the other discussion about whether a scalar default makes sense. Virtually all of the `Expr` methods assume scalars, so it's very reasonable for `Expr`'s default `_eval_derivative` to assume a scalar output (i.e., `1`). 

This doesn't necessarily need to be changed here, because like I said, it is already here, but it is something that I think should be done.

Reply:
I've altered the technique now. Instead of special-casing in `Derivative.__new__`, I've moved the "cannot compute derivative" check and return to after dummy symbols are substituted. This has the effect of giving the correct result when differentiating `_diff_wrt` objects with respect to themselves, because they now inherit the `AtomicExpr._eval_derivative_n_times` method, which has this behaviour.

Adding a default `_eval_derivative` caused the `test_diff_no_eval_derivative` test to fail, which might have been expected, but I didn't feel familiar enough with the desired semantics there to make a breaking change. I'm still not completely certain this transformation is correct (let's see what the tests say), but if it is, then it seems like a cleaner change, without introducing an extra special case where it doesn't belong.

Does this method seem cleaner to you?